File: packages/sdk/src/commands/freshness.ts
import { type Finding, finding, sortFindings } from '../lib/findings.js';
import { scanAgentProfiles, scanCategories } from '../lib/layer.js';
import { type Manifest } from '../lib/manifest.js';

export interface FreshnessItem {
   path: string;
   reviewAfter: string;
}

/** Freshness scan results: expired and upcoming review horizons. */
export interface FreshnessReport {
   /** reviewAfter date in the past. */
   expired: FreshnessItem[];
   /** reviewAfter within the next 30 days. */
   upcoming: FreshnessItem[];
   /** Total documents carrying a freshness horizon. */
   declared: number;
   findings: Finding[];
}

function reviewAfterOf(fm: Record<string, unknown> | null): string | null {
   const freshness = fm?.freshness as { reviewAfter?: unknown } | undefined;
   const v = freshness?.reviewAfter;
   return typeof v === 'string' && /^\d{4}-\d{2}-\d{2}$/.test(v) ? v : null;
}

/**
 * Report freshness horizons across category documents and agent profiles.
 * Report-only by default (warnings); --strict raises expired horizons to
 * errors. Scans documents directly so it works at any conformance level.
 */
export function freshnessReport(root: string, manifest: Manifest, strict = false): FreshnessReport {
   const today = new Date().toISOString().slice(0, 10);
   const horizon = new Date(Date.now() + 30 * 24 * 3600 * 1000).toISOString().slice(0, 10);

   const items: FreshnessItem[] = [];
   for (const doc of scanCategories(root, manifest)) {
      const reviewAfter = reviewAfterOf(doc.frontmatter);
      if (reviewAfter) items.push({ path: doc.relPath, reviewAfter });
   }
   for (const profile of scanAgentProfiles(root, manifest)) {
      const reviewAfter = reviewAfterOf(profile.frontmatter);
      if (reviewAfter) items.push({ path: profile.relPath, reviewAfter });
   }
   items.sort((a, b) =>
      a.reviewAfter !== b.reviewAfter
         ? a.reviewAfter < b.reviewAfter
            ? -1
            : 1
         : a.path < b.path
           ? -1
           : a.path > b.path
             ? 1
             : 0,
   );

   const expired = items.filter((i) => i.reviewAfter < today);
   const upcoming = items.filter((i) => i.reviewAfter >= today && i.reviewAfter <= horizon);
   const findings = expired.map((i) =>
      finding('freshness-expired', strict ? 'error' : 'warning', `review horizon ${i.reviewAfter} has passed`, i.path),
   );
   return { expired, upcoming, declared: items.length, findings: sortFindings(findings) };
}

File: packages/sdk-py/src/leji/freshness.py
"""Freshness horizons report (reviewAfter), mirroring the Node SDK."""

from __future__ import annotations

import datetime as dt
import re
from dataclasses import dataclass

from .findings import Finding, sort_findings
from .layer import scan_agent_profiles, scan_categories
from .manifest import Manifest

_DATE = re.compile(r"^\d{4}-\d{2}-\d{2}$")


@dataclass
class FreshnessReport:
    expired: list[dict]
    upcoming: list[dict]
    declared: int
    findings: list[Finding]


def _review_after_of(fm: dict | None) -> str | None:
    freshness = (fm or {}).get("freshness")
    if not isinstance(freshness, dict):
        return None
    v = freshness.get("reviewAfter")
    return v if isinstance(v, str) and _DATE.fullmatch(v) else None


def freshness_report(root: str, manifest: Manifest, strict: bool = False) -> FreshnessReport:
    """Report-only by default (warnings); --strict raises expired horizons to
    errors. Scans documents directly so it works at any conformance level."""
    today = dt.datetime.now(dt.timezone.utc).date().isoformat()
    horizon = (dt.datetime.now(dt.timezone.utc).date() + dt.timedelta(days=30)).isoformat()

    items: list[dict] = []
    for doc in scan_categories(root, manifest):
        review_after = _review_after_of(doc.frontmatter)
        if review_after:
            items.append({"path": doc.rel_path, "reviewAfter": review_after})
    for profile in scan_agent_profiles(root, manifest):
        review_after = _review_after_of(profile.frontmatter)
        if review_after:
            items.append({"path": profile.rel_path, "reviewAfter": review_after})
    items.sort(key=lambda i: (i["reviewAfter"], i["path"]))

    expired = [i for i in items if i["reviewAfter"] < today]
    upcoming = [i for i in items if today <= i["reviewAfter"] <= horizon]
    findings = [
        Finding(
            "freshness-expired",
            "error" if strict else "warning",
            f"review horizon {i['reviewAfter']} has passed",
            i["path"],
        )
        for i in expired
    ]
    return FreshnessReport(
        expired=expired, upcoming=upcoming, declared=len(items), findings=sort_findings(findings)
    )
